llvm/lib/CAS/HierarchicalTreeBuilder.cpp

```diff
@@ -257,3 +257,32 @@ Expected<TreeRef> HierarchicalTreeBuilder::create(CASDB &CAS) {
   return CAS.getTree(*Root.ID);
 }
 
+Expected<TreeRef> HierarchicalTreeBuilder::clone(CASDB &Src, TreeRef Root,
+                                                 CASDB &Dst) {
+  HierarchicalTreeBuilder Builder;
```

Comment:
A bit funny to use a tree builder here, since the builder has to do a bunch of work to figure out how things are nested after the fact.

I think this is a better framework for cloning:
- Create a map from old to new ID.
- Do a topological sort of the DAG rooted by the input (arbitrary object, not necessarily a tree), using a DFS.
- Build objects using the resulting post-order traversal (POT).

We can add peepholes to clone leaf objects during the DFS to avoid having to visit them twice. But that could land later.

Reply:
(I don't think that ends up being very complicated code...)